private/multiset.rkt

```diff
@@ -90,6 +91,22 @@
      (constructor:multiset (sub1 (multiset-size set))
                            (hash-update (multiset-frequencies set) v sub1))]))
 
+(define (multiset-union a b)
+  (define a-freqs (multiset-frequencies a))
+  (define b-freqs (multiset-frequencies b))
+  (define vals (set-union (list->set (hash-keys a-freqs))
+                          (list->set (hash-keys b-freqs))))
+  (define union-freqs
+    (for/hash ([v (in-set vals)])
+      (values v (+ (hash-ref a-freqs v 0) (hash-ref b-freqs v 0)))))
+  (constructor:multiset (+ (multiset-size a) (multiset-size b)) union-freqs))
+
+(define (multiset-add-all set seq)
+  (multiset-union set
+                  (if (multiset? seq)
```

Comment:
I realize now that the signature for `multiset-add-all` needs to chaperone the sequence with `(sequence/c any/c)`, which I think means it will always be wrapped and won't look like a multiset even if it actually is one.  So `(multiset? seq)` will always be false and this fast-path optimization won't work. I think we'll have to just repeatedly call `multiset-add` instead. On the bright side, that approach will conflict less with #248.

Reply:
A contract `(or/c multiset? (sequence/c any/c))` would ensure that `multiset`s would not be wrapped.

Reply:
That's a great idea! Let's go with that one.